pkg/fsp/header.go

```diff
@@ -0,0 +1,231 @@
+	s += fmt.Sprintf("TempRAMInit Entry Offset    : %#08x %d\n", ih.TempRAMInitEntryOffset, ih.TempRAMInitEntryOffset)
+	s += fmt.Sprintf("Reserved3                   : %#08x\n", ih.Reserved3)
+	s += fmt.Sprintf("NotifyPhase Entry Offset    : %#08x %d\n", ih.NotifyPhaseEntryOffset, ih.NotifyPhaseEntryOffset)
+	s += fmt.Sprintf("FSPMemoryInit Entry Offset  : %#08x %d\n", ih.FSPMemoryInitEntryOffset, ih.FSPMemoryInitEntryOffset)
+	s += fmt.Sprintf("TempRAMExit Entry Offset    : %#08x %d\n", ih.TempRAMExitEntryOffset, ih.TempRAMExitEntryOffset)
+	s += fmt.Sprintf("FSPSiliconInit Entry Offset : %#08x %d\n", ih.FSPSiliconInitEntryOffset, ih.FSPSiliconInitEntryOffset)
+	return s
+}
+
+// ImageRevision is the image revision fielf of the FSP info header.
+type ImageRevision uint32
+
+func (ir ImageRevision) String() string {
+	return fmt.Sprintf("%d.%d.%d.%d",
+		(ir>>24)&0xff,
+		(ir>>16)&0xff,
+		(ir>>8)&0xff,
+		ir&0xff,
+	)
+}
+
+// SpecVersion represents the spec version as a packed BCD two-digit,
+// dot-separated unsigned integer.
+type SpecVersion uint8
+
+func (sv SpecVersion) String() string {
+	return fmt.Sprintf("%d.%d", (sv>>4)&0x0f, sv&0x0f)
+}
+
+// ImageAttribute represents the image attributes.
+type ImageAttribute uint16
+
+func (ia ImageAttribute) String() string {
+	ret := fmt.Sprintf("%#04x GraphicsDisplaySupported", uint16(ia))
+	if uint16(ia) & ^(uint16(1)) != 0 {
+		ret += " (reserved bits are not zeroed)"
+	}
+	return ret
+}
+
+// ComponentAttribute represents the component attribute.
+type ComponentAttribute uint16
+
+func (ca ComponentAttribute) String() string {
+	var attrs []string
+	if uint16(ca)&0x01 == 1 {
+		attrs = append(attrs, "ReleaseBuild")
+	} else {
+		attrs = append(attrs, "DebugBuild")
+	}
+	if uint16(ca)&0x03 == 1 {
+		attrs = append(attrs, "OfficialRelease")
+	} else {
+		attrs = append(attrs, "TestRelease")
+	}
+	switch uint16(ca) >> 12 {
+	case 1:
+		attrs = append(attrs, "TypeFSP-T")
+	case 2:
+		attrs = append(attrs, "TypeFSP-M")
+	case 3:
+		attrs = append(attrs, "TypeFSP-S")
+	case 8:
+		attrs = append(attrs, "TypeFSP-O")
+	default:
+		attrs = append(attrs, "TypeFSPReserved")
+	}
+	ret := fmt.Sprintf("%#04x %s", uint16(ca), strings.Join(attrs, "|"))
+	// bits 2:12 are reserved
+	if uint16(ca)&0x0ffe != 0 {
+		ret += " (reserved bits are not zeroed)"
+	}
+	return ret
+}
+
+// FromBytes parses an FSP_INFO_HEADER from a byte buffer.
+func FromBytes(b []byte) (*InfoHeaderRev3, error) {
+	if len(b) < CommonHeaderLength {
+		return nil, fmt.Errorf("short FSP Info Header length %d; want at least %d", len(b), CommonHeaderLength)
+	}
+	var f InfoHeaderRev3
+	// parse the common header first
+	buf := uio.NewLittleEndianBuffer(b[:CommonHeaderLength])
+	f.Signature[0] = buf.Read8()
+	f.Signature[1] = buf.Read8()
+	f.Signature[2] = buf.Read8()
+	f.Signature[3] = buf.Read8()
+	f.HeaderLength = buf.Read32()
+	f.Reserved1[0] = buf.Read8()
+	f.Reserved1[1] = buf.Read8()
+	f.SpecVersion = SpecVersion(buf.Read8())
```

Comment:
The advantage of using binary.Read is that since SpecVersion is a uint8, you don't need to cast it.